### ISSUE TITLE
[New Python Track Docs]: Reading and Troubleshooting Stack Traces in Python

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3054]
This issue describes how to implement the `Stack Traces and Troubleshooting` Python Track docs.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [Existing Python Track Documents](https://exercism.org/docs/tracks/python)
- [Python Track Documents on GitHub](https://github.com/exercism/python/tree/main/docs)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This document is intended to explain basic techniques for how to read a Python traceback or `stacktrace` as well as general troubleshooting steps and (possibly) tools.  It should explain to students how to interpret Python stacktraces in general, and possibly outline common error/traceback scenarios (e.g., import errors, print errors, recursion errors, and method errors).
 
<br>

## Learning objectives

- Understand what a `stacktrace/traceback` is, and from whence it originates (_the call stack_).
- Understand where to start reading the traceback (_at the bottom_), and how each "level" up relates to the one below.
- Relate various common errors to their mistakes in code.
- Learn a little about [`traceback`](https://docs.python.org/3/library/traceback.html) and [`logging`](https://docs.python.org/3.9/library/logging.html) (_and possibly the [logging HOWTO](https://docs.python.org/3.9/howto/logging.html) and [logging cookbook](https://docs.python.org/3.9/howto/logging-cookbook.html)_), the modules that produce and log stacktraces in Python.
- Learn a little around using `print()` as a vehicle for debugging.
- Learn a little around error tracing tools such as [`PDB`](https://docs.python.org/3.9/library/pdb.html), and similar tools in common IDEs.
- Learn a little around the [`assert`](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) statement, and its role in debugging.

<br>

## Out of scope

This is a fairly high-level and broad document, emphasizing techniques for troubleshooting and debugging.  The intent here is to explain otherwise intimidating and hard to interpret error "stacks" or tracebacks in Python and give steps or strategies for both understanding and resolving them. No topic is really "out of scope", as long as it applies to stacktraces, common errors, general troubleshooting, or debugging in Python.

<br>
  
## Concepts

<details>
<summary><b>Proposed Concepts to Cover</b></summary>

- `call stacks`
-  Reading `stacktraces` or `backtraces`
- `common errors` in Python code, and their corresponding `stacktraces`
-  general/common debugging steps (_e.g. `print()` for variable values, isolating functions, stepping through code_)
-  `PDB` and other debugging tools.
- `traceback` module
- `logging` module
-  `assert` as a debugging technique.

</details>

<br>

## Prerequisites

Since this is a broad track-focused document, there really are no prerequisites.

<br>  

## Resources to refer to

<details>
<summary><b>Articles on Reading Stack Traces and Debugging in Python</b></summary>

- [Python Morsels:  Reading Tracebacks in Python](https://www.pythonmorsels.com/reading-tracebacks-in-python/)
- [Real Python: Understanding Python Tracebacks](https://realpython.com/python-traceback/)
- [Sentinel One: Python Stack Trace: Understanding and using to Debug](https://www.sentinelone.com/blog/python-stack-trace-understanding-using-debug/)
- [Wikipedia:  Stack Traces]( https://en.wikipedia.org/wiki/Stack_trace)
- [PyMOTW-3: traceback - the module for Exceptions and Stack Traces](http://pymotw.com/3/traceback/)
- [Python Docs: `traceback` module](https://docs.python.org/3/library/traceback.html)
- [Python Docs: `sys.excepthook()` and friends](https://docs.python.org/3/library/sys.html#sys.excepthook)
- [Real Python: Python Logging](https://realpython.com/python-logging-source-code/)
- [Python Docs: `logging` module](https://docs.python.org/3.9/library/logging.html)
- [CS Franklin: How to Read a Traceback](http://cs.franklin.edu/~ansaria/traceback.html)
- [Real Python: Python's Assert](https://realpython.com/python-assert-statement/)
- [Python Docs: `assert`](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement)
- [Real Python: Debugging with PDB](https://realpython.com/python-debugging-pdb/)
- [Python Docs: `PDB`](https://docs.python.org/3.9/library/pdb.html)
- [Automation Panda: The Testing Pyramid](https://automationpanda.com/2018/08/01/the-testing-pyramid/)
- [Automation Panda: Testing]( https://automationpanda.com/testing/)

</details>
<br>

### Track Document

Since this is a track document, it can be formatted in markdown any way the author desires.  However, it should still conform to the [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown), and follow all the rules in the [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide).  Any links should be reference links, and any included images should be placed in the `docs/images` folder.

### Document Entry in `docs/config.json`

See other entries in the [`config`](https://github.com/exercism/python/blob/main/docs/config.json) doc.  Note that each document displayed on the website needs:

- [ ] A valid UUID v4 number.  You can use `configlet`, or [this generator](https://www.uuidgenerator.net/version4).
- [ ] A slug for the document name
- [ ] A path (_typically `docs/<DOCUMENTNAME>.md`_)
- [ ] A title
- [ ] A blurb of less than 350 characters, describing what the document is/explains.

<br>

## Implementation Notes

- Tone should be friendly and semi formal.  See other track docs for a general guide
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

## Help

If you have any questions while implementing this issue, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
Closing, since #3059 has been merged.
